### PR TITLE
[DEV-2939] Integrate sumsub

### PR DESCRIPTION
**Pull request by Kolibri1990**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+      '325528857': {
+        reviewAnswer: 'GREEN';
+      };
+    };
+    forbidden: false;
+    partialCompletion: null;
+    stepStatuses: null;
+    imageStatuses: [];
+  };
+}
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/sum-sub.dto.ts:69]: Fix types (remove fixed string values, nulls, booleans, ...)

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+export enum ReviewAnswer {
+  GREEN = 'green',
+  RED = 'red',
+}
+
+export enum ReviewRejectType {
+  FINAL = 'final',
+  RETRY = 'retry',
+}
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/sum-sub.dto.ts:89]: Looks a bit unfinished 🤔  (e.g. IdDocType)

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+export enum ApplicantType {
+  COMPANY = 'company',
+  INDIVIDUAL = 'individual',
+}
+
+export enum ReviewAnswer {
+  GREEN = 'green',
+  RED = 'red',
+}
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/sum-sub.dto.ts:84]: Is it `'green'` or `'GREEN'`?

File: src/subdomains/generic/kyc/controllers/kyc.controller.ts [hunk 4]
```diff
@@ -245,6 +248,22 @@ export class KycController {
     }
   }
 
+  @Post('sumSub')
+  @ApiExcludeEndpoint()
+  async sumSubWebhook(@Req() req: Request, @Body() data: SumSubResult) {
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/controllers/kyc.controller.ts:253]: `@Post('ident/sumSub')` and move the endpoint above `ident/:type`

File: src/subdomains/generic/kyc/services/integration/sum-sub.service.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      'GET',
+      JSON.stringify({}),
+      'arraybuffer',
+    )
+      .then(Buffer.from)
+      .catch((e) => {
+        this.logger.error(`Failed to fetch ${applicantType} summary report from ${applicantId}:`, e);
+        return undefined;
+      });
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/sum-sub.service.ts:50]: Move the `JSON.stringify` into the `callApi` method?

File: src/subdomains/generic/kyc/services/integration/sum-sub.service.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+@Injectable()
+export class SumSubService {
+  private readonly logger = new DfxLogger(SumSubService);
+
+  private readonly baseUrl = `https://api.sumsub.com`;
+
+  constructor(private readonly http: HttpService) {}
+
+  async createApplicant(user: UserData, kycStep: KycStep, kycLevel: string): Promise<void> {
+    await this.callApi<{ id: string }>(
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/sum-sub.service.ts:21]: What is this `kycLevel` string?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 7]
```diff
@@ -749,4 +784,16 @@ export class KycService {
       );
     }
   }
+
+  private async downloadSumSubDocuments(user: UserData, namePrefix = '') {
+    const data = await this.sumSubService.getDocument('669a1dd87953b83e1ae80426', ApplicantType.INDIVIDUAL);
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:789]: Fancy number?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 8]
```diff
@@ -392,6 +395,38 @@ export class KycService {
...
+      case IdentShortResult.SUCCESS:
+        user = user.internalReviewStep(kycStep, dto);
+        await this.downloadSumSubDocuments(user);
+        break;
+
+      case IdentShortResult.FAIL:
+        user = user.failStep(kycStep, dto);
+        await this.downloadSumSubDocuments(user, 'fail/');
+        //await this.kycNotificationService.identFailed(user, getIdentReason(reason)); TODO: Create notifications
+        break;
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:420]: TODO?

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts (hunk 1 above)
Comment (Kolibri1990) [reply src/subdomains/generic/kyc/dto/sum-sub.dto.ts:69]: Is not in use. Will remove it was from some testing.

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts (hunk 2 above)
Comment (Kolibri1990) [reply src/subdomains/generic/kyc/dto/sum-sub.dto.ts:89]: Is also not in use.

Comment (davidleomay) [reply src/subdomains/generic/kyc/dto/sum-sub.dto.ts:89]: Ok 👍

File: src/subdomains/generic/kyc/dto/sum-sub.dto.ts (hunk 1 above)
Comment (davidleomay) [reply src/subdomains/generic/kyc/dto/sum-sub.dto.ts:69]: Ok 👍

**Review by davidleomay — APPROVED** [review]

**Pull request merged**